### ISSUE TITLE
Use custom TravisCI matrix?

### ISSUE THREAD

**Issue by eventualbuddha** [decaffeinate/decaffeinate-examples#7]
I think we should set this project up such that each example is run as a separate Travis worker, and that the ones that don't yet work are marked as allowed failures. What do you think?

**Comment by alangpierce** [comment]
Yeah, certainly worth discussion. It's hours of compute time to run this for all projects, so I'd rather not run the full suite on every PR/commit to this repo. My thought was to have a weekly build or something like that, which I think can be done with the cron job system in travis.

For individual commits to the repo, my plan was to just do more of a smoke test, currently just converting the hubot repo, but maybe at some point there will also be unit tests and things like that. I guess it also would be nice if the build for PRs/commits in this repo didn't break due to regressions in decaffeinate or changes in these projects.

My vision for the weekly build was something like this:
* Rather than cloning the upstream projects directly, it clones a fork in the decaffeinate-examples org (which doesn't exist yet).
* It also updates the `master` branch for that fork to pull in any upstream changes from the last week.
* It converts all files that can be converted and runs tests and creates a report of the results.
* It overwrites the README in the repo with a presentable explanation of the results (e.g. "58/63 files successfully converted and 600/800 tests passed"). The decaffeinate-errors.log file is committed to the repo and linked from the README, and maybe also a transcript of the test output (in both cases, only if failures actually happened).
* The converted repo with a description of the conversion status is force-pushed to a `decaffeinate` branch (maybe with a new tag each time), which is the default branch in the fork. That way, anyone visiting the github project can see the current conversion status and explore the JS code and see what errors happened.
* The README for this repo has a table of the high-level status of all projects, and it gets updated as part of the weekly build.

IMO, having all of the information on github rather than travis makes the information a lot more accessible. But automatic alerting for regressions is nice as well, of course. I guess maybe one thought there is to do the build matrix thing like you mentioned, but still update the repo even if there are failures (and mark the build as failed). That should make unexpected decaffeinate failures louder, but one downside is it may be hard to distinguish between a failure due to a decaffeinate regression and a failure due to the tool not working.

**Comment by eventualbuddha** [comment]
I do like the idea of surfacing the information in the repo itself. Maybe having some way of noticing decreasing test passage?

**Comment by alangpierce** [comment]
Ok, I got a weekly cron job set up. Here's how it's all hooked up:

Since I want two separate "types" of builds (one that just makes sure this repo is working, and one that actually does a bunch of concurrent publishes), I decided to just make a new repo for the cron job. It's at https://github.com/decaffeinate/decaffeinate-example-builder . (Another option was a different branch, but I find it annoying to have two dramatically different branches in the same repo. Another option would be to have the .travis.yml check an environment variable to see if it's running from a cron job, and do something different if so, but that also seemed annoying, particularly since one type of build should use a build matrix and the other shouldn't.)

Since I need the script to push to a bunch of repos, I needed to figure out credentials. To help avoid dumb mistakes, I made a bot account rather than the git operations running under my own user (and I think that also looks a little nicer).

I made a Google account, decaffeinatebot@gmail.com, and a GitHub account, https://github.com/decaffeinate-bot, and I added decaffeinate-bot to the decaffeinate-examples organization with write access to all repos. It isn't part of the decaffeinate org since (currently) it doesn't need write access to anything there. The passwords for the Google and GitHub account were randomly generated and are in a LastPass account that I have, and I can send them to you if you ever want.

There were various ways to set up git auth on Travis, but I ended up going with SSH auth for the whole user account. (GitHub lets you specify repo-specific deploy keys, but apparently every repo needs a different key, which is annoying for this case.) I generated a public/private ssh key pair and used `travis encrypt-file` on the private key so that the encrypted file is in the repo and the decryption key is an environment variable. I then unencrypt it into `~/.ssh/id_rsa`

If you're curious, the details are all in https://github.com/decaffeinate/decaffeinate-example-builder/blob/master/.travis.yml

I then went into the travis config and scheduled a weekly cron job using their relatively new cron feature.

It actually only takes about 20 minutes of clock time for the job to run (the slowest by far is codecombat), which is faster than I thought it would be. So probably it could be a daily job if we want.

Still plenty more that could be done (e.g. detecting regressions, like you mentioned), but I think all the basic stuff is in place now.
